lib/Sema/CSDiag.cpp

```diff
@@ -5291,7 +5310,18 @@ bool FailureDiagnosis::visitApplyExpr(ApplyExpr *callExpr) {
   // Collect a full candidate list of callees based on the partially type
   // checked function.
   CalleeCandidateInfo calleeInfo(fnExpr, hasTrailingClosure, CS);
-  
+
+  // Filter list of the candidates based on the known function type.
+  if (auto fn = fnType->getAs<AnyFunctionType>()) {
+    using Closeness = CalleeCandidateInfo::ClosenessResultTy;
+    calleeInfo.filterList([&](UncurriedCandidate candidate) -> Closeness {
+      auto resultType = candidate.getResultType();
+      if (resultType && !resultType->isEqual(fn->getResult()))
```

Comment:
This seems like it's always going to filter out generic candidates since the result type will not be equal.

I'm wondering if there is an easy way to improve that. Offhand I don't know.